tao_compiler/mlir/disc/transforms/codegen_utils.cc

```diff
@@ -390,5 +392,296 @@ int getReductionTileSizeOnCPU() {
+    if (op.getNumRegions() != 0) {
+      return success();
+    }
+  }
+
+  // Check that the number of cloned instructions is correct.
+  int64_t numEffectiveOps = std::distance(loopBodyBlock->begin(),
+                                          std::prev(loopBodyBlock->end(), 1)) -
+                            beginIdx;
+  if (numEffectiveOps != numOps * unrollFactor) {
+    return failure();
+  }
+
+  // Make sure `toMove` is after `toMoveAfter` when calling this lambda
+  // function.
+  auto canMoveAfter = [&](Block::iterator toMove, Block::iterator toMoveAfter) {
+    assert(toMove->getParent() == toMoveAfter->getParent());
+    // DominanceInfo dom(toMove->getParentOp());
+    bool canMove = true;
+    for (auto iter = std::prev(toMove); iter != toMoveAfter;
+         std::advance(iter, -1)) {
+      // if (dom.dominates(&(*iter), &(*toMove))) {
+      if (dependsOn(&(*toMove), &(*iter))) {
+        canMove = false;
+        break;
+      }
+    }
+    return canMove;
+  };
+
+  bool canMove = true;
+  // Note only the first `numOps - 1` ops need to be moved to do interleaving.
+  for (int64_t opIdx = 0; opIdx < (numOps - 1) && canMove; opIdx++) {
+    auto firstInst =
+        std::next(loopBodyBlock->begin(), beginIdx + opIdx * unrollFactor);
+    auto toMoveAfter = firstInst;
+    for (int64_t unrollIdx = 1; unrollIdx < unrollFactor && canMove;
+         unrollIdx++) {
+      auto toInterleave = std::next(firstInst, (numOps - opIdx) * unrollIdx);
+      canMove &= canMoveAfter(toInterleave, toMoveAfter);
+      if (canMove) {
+        toInterleave->moveAfter(&(*toMoveAfter));
+        std::advance(toMoveAfter, 1);
+      }
+    }
+  }
+
+  return success();
+}
+
+// This function unrolls the given for-loop op and interleaves the instructions
+LogicalResult loopUnrollByFactorAndTryInterleave(
+    scf::ForOp forOp, uint64_t unrollFactor,
+    function_ref<void(unsigned, Operation*, OpBuilder)> annotateFn) {
+  assert(unrollFactor > 0 && "expected positive unroll factor");
+
+  // Return if the loop body is empty.
+  if (llvm::hasSingleElement(forOp.getBody()->getOperations()))
+    return success();
+
+  // Compute tripCount = ceilDiv((upperBound - lowerBound), step) and populate
+  // 'upperBoundUnrolled' and 'stepUnrolled' for static and dynamic cases.
+  OpBuilder boundsBuilder(forOp);
+  auto loc = forOp.getLoc();
+  auto step = forOp.getStep();
+  Value upperBoundUnrolled;
+  Value stepUnrolled;
+  bool generateEpilogueLoop = true;
+
+  auto lbCstOp = forOp.getLowerBound().getDefiningOp<arith::ConstantIndexOp>();
+  auto ubCstOp = forOp.getUpperBound().getDefiningOp<arith::ConstantIndexOp>();
+  auto stepCstOp = forOp.getStep().getDefiningOp<arith::ConstantIndexOp>();
+  if (lbCstOp && ubCstOp && stepCstOp) {
+    // Constant loop bounds computation.
+    int64_t lbCst = lbCstOp.value();
+    int64_t ubCst = ubCstOp.value();
+    int64_t stepCst = stepCstOp.value();
+    assert(lbCst >= 0 && ubCst >= 0 && stepCst >= 0 &&
```

Comment:
A rookie's question, why only support positive loop bounds, it seems 
`for i = -10 to -1` also works well from this code segment?

Reply:
This is to ease the programming. There is no scenario to optimize loops with negative ub/lb to unroll for optimizing. Only dealing with positive ub/lb is enough for the optimization.